### PR TITLE
avahi: new formula (Linux-only)

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by danielnachun** [comment]
I guess this needs https://github.com/Homebrew/homebrew-core/pull/75915 to be merged first.

File: Formula/avahi.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                             "--disable-gtk",
+                             "--disable-gtk3",
+                             "--disable-libevent",
+                             "--with-distro=none",
+                             "--with-systemdsystemunitdir=no"
+    system "make", "install"
+  end
+
+  test do
+    assert_match "avahi-browse #{version}", shell_output("#{bin}/avahi-browse --version").chomp
```
Comment (SMillerDev) [inline Formula/avahi.rb:42]: Nothing else we can test?

Comment (danielnachun) [reply Formula/avahi.rb:42]: Agreed, there should be something better.  I'll look for some examples that will work in Docker.

**Review by MikeMcQuaid — DISMISSED** [review]

**Comment by danielnachun** [comment]
Does anyone know why all the dependencies for this are now being built from source in Linux CI?  I thought this wasn't supposed to happen?  I'm also seeing the same thing happen for https://github.com/Homebrew/homebrew-core/pull/76059.

File: Formula/avahi.rb (hunk 1 above)
Comment (danielnachun) [reply Formula/avahi.rb]: I wrote a much more extensive test now that includes some headers and links to several of the shared libraries.

**Comment by iMichka** [comment]
Because in homebrew-core there are no bottles, so everything is built from source for Linux.

**Comment by danielnachun** [comment]
> Because in homebrew-core there are no bottles, so everything is built from source for Linux.

Indeed, not sure how that slipped my mind.  Now I get to search through the needle in the haystack for what failed.  I'm assuming there is some dependency that hasn't been backported to work on Linux.

**Comment by iMichka** [comment]
There is no clear failure, I would be that something wrong happened in the perl postinstall block.

**Comment by danielnachun** [comment]
Yeah that would make sense.  There were 8 formulae where the output was truncated with "Only the first 100 operations were put." which could be hiding a post install failure.  Is there way to recover more complete logs?  Maybe we could open a "do nothing" PR (just add a blank line) for Perl and force it to be tested in Linux CI here?

**Comment by danielnachun** [comment]
I'm stumped with this one.  There don't appear to be any artifacts generated, perhaps because it's Linux only?  Definitely open to any suggestions on how to move this forward.

**Comment by iMichka** [comment]
#76753 should help

**Comment by danielnachun** [comment]
> #76753 should help

That looks awesome.  I'll rerun these jobs after it gets merged.

**Comment by iMichka** [comment]
You can now rebase this.

**Comment by danielnachun** [comment]
After the rebase I'm getting this error:

`No files were found with the provided path: ~/bottles/logs. No artifacts will be uploaded.`

I think I rebased to include the update, but maybe I did something wrong?

**Comment by iMichka** [comment]
Maybe it's because HOMEBREW_DEVELOPER was not set. Here is another attempt to fix this: #76806

**Comment by iMichka** [comment]
The issue is test-bot: https://github.com/Homebrew/homebrew-test-bot/blob/8c327484819bf6f8b0212cea0abd3eb7ed780abd/cmd/test-bot.rb#L90

Maybe we could try to set the --local flag in the CI file.

**Comment by danielnachun** [comment]
Adding `--local` to the `brew test-bot` call in the CI file did finally let us upload the log files.  But now that I can see them they seem to be just as inconclusive as the other log files we already have.  This a very mysterious failure!

**Comment by iMichka** [comment]
At this point I would bet on a failure in a postinstall block. It's really annoying that these do not appear in a clearer way in the logs. Perl looks suspicious too.

You might open a second PR in this repo and test Perl alone for example.

**Comment by danielnachun** [comment]
> At this point I would bet on a failure in a postinstall block. It's really annoying that these do not appear in a clearer way in the logs. Perl looks suspicious too.
> 
> You might open a second PR in this repo and test Perl alone for example.

I was thinking that it was about time that I just test all the dependencies alone in homebrew-core and find which one fails.  What's odd here is that among the logs from the build artifact, `avahi` appears to build just fine.  So I agree that a post install failure is most likely, as the dependencies would still be installed.

For making these PRs, I think I should add them to `linux-pr-tests.yml` and also set `depends_on :linux` to avoid wasting CI resources on macOS, since the PRs won't actually be merged.  Does that sound like the best approach?  I'll also give a heads up on Slack before I open those tomorrow, and tag all of the them with `do-not-merge`.

**Comment by MikeMcQuaid** [comment]
> At this point I would bet on a failure in a postinstall block. It's really annoying that these do not appear in a clearer way in the logs. Perl looks suspicious too.

@iMichka How would you like them to be displayed compared to how they are currently displayed?

**Comment by iMichka** [comment]
So right now I can grep for "failed" or "error" in the logs, but I can't see where the failure is. It tells me "Error: 1 failed step!", but I don't which step in the process exactly failed.

**Comment by MikeMcQuaid** [comment]
> So right now I can grep for "failed" or "error" in the logs, but I can't see where the failure is. It tells me "Error: 1 failed step!", but I don't which step in the process exactly failed.

@iMichka Can you link to a specific log and what you're looking for in it? Thanks!

**Comment by iMichka** [comment]
> @iMichka Can you link to a specific log and what you're looking for in it? Thanks!

Yes, it is in this PR, it is the only test that failed (GitHub Actions CI - Linux / tests_linux (pull_request)). We can not find out from the output what exactly failed in the whole build. One of the substeps failed, but which one ...

**Comment by danielnachun** [comment]
A direct link to the failing log is: https://github.com/Homebrew/homebrew-core/commit/1829095682b8aa447e41997d91df1ca5173e6d6e/checks/2541073987/logs.

**Comment by MikeMcQuaid** [comment]
> Yes, it is in this PR, it is the only test that failed (GitHub Actions CI - Linux / tests_linux (pull_request)). We can not find out from the output what exactly failed in the whole build. One of the substeps failed, but which one ...

`brew install --only-dependencies --verbose avahi`. That's what `Output brew test-bot --only-json-tab --only-formulae failures` says. Is that specific enough? "Run mkdir ~/bottles" definitely looks wrong for a set name.

**Comment by danielnachun** [comment]
The problem is that we don't know which dependency failed, as there are no identifiable failures in the logs for the installation of those dependencies.  The tricky thing here is that have to build all the dependencies from source since we don't currently have Linux bottles in homebrew-core.

**Comment by MikeMcQuaid** [comment]
> The problem is that we don't know which dependency failed, as there are no identifiable failures in the logs for the installation of those dependencies. The tricky thing here is that have to build all the dependencies from source since we don't currently have Linux bottles in homebrew-core.

Yes, gotcha. That's non-ideal, I agree. Given it's very specific to the current homebrew-core Linux situation I'm hesitant to put time into it myself but hopefully will get addressed by the merge 🤞🏻

**Comment by iMichka** [comment]
@danielnachun I would recommend opening a linux build for perl in this repo. We need to have a green CI here to be able to proceed.

**Comment by danielnachun** [comment]
I think I know what the problem might be.  `dbus` depends on `xmlto` which depends on `gnu-getopt`.  `glib` depends on `util-linux`, and `gnu-getopt` and `util-linux` install conflicting files. I will upstream the fix to make `gnu-getopt` keg-only on Linux, and rebase this after merging that fix.

**Comment by danielnachun** [comment]
I added the fix to make `gnu-getopt` keg-only and it worked!  

@MikeMcQuaid @iMichka I guess the issue here was that `brew link` failed for either `gnu-getopt` or `util-linux` due to conflicting files, and for some reason that was not printed to the logs.  In the linuxbrew-core CI workflow, when `brew link` fails it does print an error after trying to link the conflicting dependency.  

I'm assuming the same thing happens in homebrew-core on macOS, so maybe this is just a quirk of the Linux CI set up here?  I haven't had a PR in homebrew-core yet where dependencies had a file conflict that made `brew link` fail so I don't know if the output is the same as in linuxbrew-core.

**Comment by iMichka** [comment]
Can we move the keg_only fix to a separate PR?

I also think we might need some improvements around the keg_only part, maybe a new syntax:
```
keg_only :provided_by_macos, "conflicts with gnu-getopt"
```

Or maybe allow to put a second keg_only inside an `on_linux` block.

**Comment by danielnachun** [comment]
I already have a separate PR for `gnu-getopt` here: https://github.com/Homebrew/homebrew-core/pull/77336.  I agree that new syntax you proposed makes sense.  I will open issue to get some feedback on how best to handle this.

**Comment by danielnachun** [comment]
I think this is ready to merge now, as the CI is all green!

**Review by SMillerDev — DISMISSED** [review]

**Pull request merged**